app/prometheus_unix.ml

```diff
@@ -42,17 +42,41 @@ module Unix_runtime = struct
   ]
 end
 
-type config = int option
+type config = {
+  port : int option;
+  addr : string option;
+}
 
 module Server = Prometheus_app.Cohttp(Cohttp_lwt_unix.Server)
 
-let serve = function
-  | None -> []
-  | Some port ->
+let serve config = match config.port, config.addr with
+  | None, _ -> []
+  | Some port, None ->
     let mode = `TCP (`Port port) in
     let callback = Server.callback in
     let thread = Cohttp_lwt_unix.Server.create ~mode (Cohttp_lwt_unix.Server.make ~callback ()) in
     [thread]
+  | Some port, Some addr  ->
+    let open! Unix in
+    let [@ocaml.warning "-partial-match"] addrinfo :: _ =
+      getaddrinfo addr (Int.to_string port) [AI_SOCKTYPE SOCK_STREAM] in
+    let socket = Lwt_unix.socket ~cloexec:true addrinfo.ai_family addrinfo.ai_socktype addrinfo.ai_protocol in
+    let () = Lwt_unix.setsockopt socket SO_REUSEADDR true in
+    let mode = `TCP (`Socket socket) in
+    let callback = Server.callback in
+    let () = Lwt.async (fun () -> Lwt_unix.bind socket addrinfo.ai_addr) in
```

Comment:
This isn't going to be reliable. It should be done inside `thread`, not racing with it.

Reply:
Made the port binding synchronous, this should be a lot more reliable.